Data loss protection

halseth opened the pull request:
This PR implements the logic required for "data loss protection", as described in BOLT#2: https://github.com/lightningnetwork/lightning-rfc/blob/master/02-peer-protocol.md#message-retransmission

When syncing channel states we will now detect the cases where _we_ probably have lost our local state, meaning broadcasting our commitment would be unsafe as it could be considered a breach. Instead we store the `my_current_per_commitment_point` sent to us by the remote in the database. This commitment point can later be used to reclaim our funds if the remote party decides to unilaterally close the channel using the corresponding state.

If we detect that the remote party probably has lost state, we'll be a good citizen and force close the channel using our latest commitment.

@Roasbeef: what is meant by:
> In order to ensure we can carry out this process reliably we may need to ensure that they remote party first sends their re-sync message before we do. We can enforce this by requiring the initiator to send their message first.

?

Fixes #1131

irekzielinski commented:
Can't wait for this to be merged in - this will allow BLW wallet to use LND nodes!
Thank you for your effort guys!

lcasassa commented:
Any updates on this?

Roasbeef commented:
I have a pending review which should be posted soon.

On Wed, Jul 4, 2018, 2:27 PM Olaoluwa Osuntokun <laolu32@gmail.com> wrote:

> You're free to help test and review Linus
>
> On Wed, Jul 4, 2018, 2:07 PM Linus Casassa <notifications@github.com>
> wrote:
>
>> Any updates on this?
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/lightningnetwork/lnd/pull/1364#issuecomment-402544627>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AA87Lh_ILbFtX8Xd9yJ8-xvYDr1C_xO2ks5uDRJygaJpZM4Uira9>
>> .
>>
>

lcasassa commented:
I'm happy to test. But I get an error when compiling the docker image after doing a rebase from master. :/ If I don't do the rebase I get the offline channel error.

Roasbeef commented on lnwallet/channel_test.go:
```diff
@@ -3486,6 +3487,228 @@ func TestChanSyncOweRevocationAndCommitForceTransition(t *testing.T) {
 	}
 }
 
+// TestChanSyncFailure tests the various scenarios during channel sync where we
```
Solid tests 👍

Roasbeef commented on htlcswitch/link.go:
```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
+				l.fail(
+					LinkFailureError{
+						code:       ErrSyncError,
+						ForceClose: true,
```
Reminder that we should also record this instance such that we re-send the chan sync message each time we connect to the remote party until they sweep the output on chain. No reason this needs to be done in this current PR as it's already size-able, so we can just add a TODO here and an issue so we can proceed w/ review+testing of this PR.

halseth replied:
Added TODO, will create an issue to address this when merged.

Roasbeef commented on htlcswitch/link.go:
```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
+				)
+				break
+
+			case err == lnwallet.ErrCommitSyncLocalDataLoss:
```
Shouldn't we signal this to the chain watcher at this point?

Roasbeef commented on htlcswitch/link.go:
```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
+				// The remote has sent us its latest unrevoked
+				// commitment point, that we stored in the
+				// database, that we can use to retrieve the
+				// funds when the remote closes the channel.
```
Style nit on these comments: typically we place them right above the `case` statement. A bonus of this is that the blocks are shorter as there's less line wrapping needed.

Roasbeef commented on contractcourt/chain_watcher.go:
```diff
@@ -379,15 +379,45 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
+			// If we are lucky, the remote peer sent us the correct
+			// commitment point during channel sync, such that we
+			// can sweep our funds.
+			commitPoint, err := c.cfg.chanState.DataLossCommitPoint()
```
commintPoint -> commitPoint 

What happens if we detect this force close before we actually write this state to disk? Since this is usually expected _before_ we connect to the remote peer it seems likely. Also consider the case where they've already swept on-chain, but we weren't online, but they keep re-broadcasting the last chan sync message so we can get the revocation point. It would seem that we need some additional synchronization in order to resolve this in all scenarios.

halseth replied:
That is a very good point! 

I briefly started working on this, but realized that this require some bigger changes:

**1) In the case we are about to receive the message, but detecting the force close in chain wins the race.** We must retry fetching the commit point after a timeout, or a signal that a new commitpoint was received.

**2) In the case where the remote has closed the channel while we have been offline.** We must support resending the channel sync message such that we can continue when it is received.

Since 2) seems to require 1), it makes sense do add them both in the follow-up adding chanSync message resend. The current implementation will handle the case where we detect a desync during channel reestablishment, and one party force closes because of this (Note that 1) is unlikely in this case, as we wait for the force close to be confirmed before we act).

Roasbeef replied:
SGTM

Roasbeef commented on contractcourt/chain_watcher.go:
```diff
@@ -379,15 +379,45 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
+					c.cfg.chanState.FundingOutpoint, err)
+				return
+			}
+			log.Infof("Recovered commit point for channel(%v)! "+
```
Style nit: missing a new line above.

Roasbeef replied:
May be worth also logging the hex encoding of the commitment point here.

Roasbeef commented on contractcourt/chain_watcher.go:
```diff
@@ -379,15 +379,45 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
+			// state, we'll just pass an empty commitment. Note
+			// that this means we won't be able to recover any HTLC
+			// funds.
+			// TODO(halseth): can we try to recover some HTLCs?
```
Maybe...only if we have _partial_ state, and this isn't actually the result of us restoring w/ seed+static-backups.

Roasbeef commented on contractcourt/chain_watcher.go:
```diff
@@ -379,15 +379,45 @@ func (c *chainWatcher) closeObserver(spendNtfn *chainntnfs.SpendEvent) {
+			// funds.
+			// TODO(halseth): can we try to recover some HTLCs?
+			err = c.dispatchRemoteForceClose(
+				commitSpend, channeldb.ChannelCommitment{},
```
Ah now I see why the `commitPoint` and chan state are passed distinctly.

Roasbeef commented on lnd_test.go:
```diff
@@ -5661,19 +5661,459 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
+	}()
+
+	// Dave will be the party losing his state.
+	dave, err := net.NewNode("Dave", []string{})
```
Can just pass a `nil` here as the second param.

Roasbeef commented on lnd_test.go:
```diff
@@ -5661,19 +5661,459 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
+
+	// We must let Dave communicate with Carol before they are able to open
+	// channel, so we connect Dave and Carol,
+	if err := net.ConnectNodes(ctxb, carol, dave); err != nil {
```
This'll fail after rebasing to master as `carol` was started in `--nolisten` mode.

Roasbeef commented on lnd_test.go:
```diff
@@ -5661,19 +5661,459 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
+	block = mineBlocks(t, net, 1)[0]
+	assertTxInBlock(t, block, daveSweep)
+
+	// Now Dave should considere the channel fully closed.
```
considere -> consider?

halseth commented on lnwallet/channel.go:
```diff
@@ -66,20 +66,40 @@ var (
+	// data, and must fail the channel and MUST NOT force close it. Instead
+	// we should wait for the remote to force close it, such that we can
+	// attempt to sweep our funds.
+	ErrCommitSyncLocalDataLoss = fmt.Errorf("possible local commitment " +
```
Added a few TODOs to address this. Makes sense to do in a follow up as this PR doesn't alter how this is already enforced in master (it is not).

Will create issue when merged.

halseth commented on lnd_test.go:
```diff
@@ -4733,16 +4733,38 @@ func testRevokedCloseRetribution(net *lntest.NetworkHarness, t *harnessTest) {
+
+	// We must let Bob communicate with Carol before they are able to open
+	// channel, so we connect Bob and Carol,
+	if err := net.ConnectNodes(ctxb, carol, net.Bob); err != nil {
```
`carol` is connecting `bob`, but `--nolisten` makes sure `bob` cannot connect to `carol`. Not sure how this worked before that bug was fixed, but it is still working it looks like.

halseth commented on lnd_test.go:
```diff
@@ -4733,16 +4733,38 @@ func testRevokedCloseRetribution(net *lntest.NetworkHarness, t *harnessTest) {
 	ctxt, _ := context.WithTimeout(ctxb, timeout)
 	chanPoint := openChannelAndAssert(
-		ctxt, t, net, net.Alice, net.Bob, chanAmt, 0, false,
+		ctxt, t, net, carol, net.Bob, chanAmt, 0, false,
```
Since we cannot set Alice to be `--nolisten`.

halseth commented on lnd_test.go:
```diff
@@ -4981,27 +5007,44 @@ func testRevokedCloseRetributionZeroValueRemoteOutput(net *lntest.NetworkHarness
 	// announcement, so we open a channel with Carol,
-	if err := net.ConnectNodes(ctxb, net.Alice, carol); err != nil {
-		t.Fatalf("unable to connect alice to carol: %v", err)
+	if err := net.ConnectNodes(ctxb, dave, carol); err != nil {
```
See other comment.

halseth commented on channeldb/channel.go:
```diff
@@ -669,6 +677,81 @@ func (c *OpenChannel) MarkAsOpen(openLoc lnwire.ShortChannelID) error {
+			return err
+		}
+
+		channel.ChanStatus = Borked
```
Agreed! Added state `LocalDataLoss`. Could add more types in the future for the different close types.

halseth commented on lnwallet/channel.go:
```diff
@@ -3311,10 +3306,52 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
+	// trick us. Since we have no way of verifying whether they are lying
+	// or not, we will fail the channel, but should not force close it
+	// automatically.
+	case msg.NextLocalCommitHeight > remoteTipHeight+1:
```
It has already been checked in the above `switch`.

halseth commented:
Rebased and addressed comments. Also increased the cases we force close to include the remote giving us invalid data.

I will create a few issues to handle the remaining follow-ups, most notably prohibiting the user from force closing a desynced channel, and adding chanSync message resend.

PTAL

Roasbeef commented:
Needs a rebase!

Roasbeef commented:
Also the follow up issues need to be created as well so we don't lose track of them.

cfromknecht commented on lnwallet/channel.go:
```diff
@@ -3258,17 +3372,51 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
+	// is valid.
+	var commitPoint *btcec.PublicKey
+	switch {
+	case msg.NextLocalCommitHeight == remoteTailHeight+2:
```
This line doesn't seem reachable, since first check in switch above is 
```
case msg.NextLocalCommitHeight > remoteTailHeight+1:
    return ErrCannotSyncCommitChains
```

halseth replied:
That's `msg.NextLocalCommitHeight > remoteTipHeight+1`. The case where we send a new commitment to the remote (new remote _tip_) that they did not receive, results in `msg.NextLocalCommitHeight == remoteTipHeight`. The case where they receive and ACK this commitment results in `msg.NextLocalCommitHeight == remoteTipHeight+1`. Higher commit heights will be invalid.

The case `msg.NextLocalCommitHeight == remoteTailHeight+2` can happen when the new remote tip is ACKed, but we didn't receive the ACK. In that case we have a pending remote commitment, meaning `remoteTipHeight == remoteTailHeight+1`.

cfromknecht replied:
ahh i see, i misread tip as tail. thanks for the explanation :)

cfromknecht commented on channeldb/channel.go:
```diff
@@ -671,6 +688,81 @@ func (c *OpenChannel) MarkAsOpen(openLoc lnwire.ShortChannelID) error {
+	}
+
+	// Update the in-memory representation to keep it in sync with the DB.
+	c.ChanStatus = LocalDataLoss
```
do we ever check that we're not overwriting another subsystem's calls?

If we make `ChannelStatus` a bitvector, we can allow the calls to not overwrite each other

halseth replied:
Good call! I made it into a bitvector, and since the existing values used were 0, 1 and 2 (`0b10`), this worked out nicely with the new one being `4`.

Also made sure that all places we are reading this value wouldn't be affected, as we are just checking that it is not "Default".

halseth commented on htlcswitch/link.go:
```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
+					"unable to synchronize channel "+
+						"states: %v", err,
+				)
+				break
```
Yes!

halseth commented on htlcswitch/link.go:
```diff
@@ -797,12 +794,61 @@ func (l *channelLink) htlcManager() {
+				fallthrough
+
+			default:
+				l.fail(
```
Done 👍

halseth commented on lnwallet/channel.go:
```diff
@@ -3259,17 +3371,44 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
 
-	} else if remoteChainTip.height+1 != msg.NextLocalCommitHeight {
+	// There should be no other possible states as long as the commit chain
+	// can have at most two elements. If there is something is wrong.
```
Reworded.

cfromknecht commented on channeldb/channel.go:
```diff
@@ -312,8 +327,10 @@ func (c ChannelStatus) String() string {
+		return "LocalDataLoss"
 	default:
-		return "Unknown"
+		return fmt.Sprintf("Unknown(%08b)", c)
```
nice formatting directive 👌

cfromknecht commented on channeldb/channel.go:
```diff
@@ -705,15 +797,15 @@ func (c *OpenChannel) putChanStatus(status ChannelStatus) error {
 
 	// Update the in-memory representation to keep it in sync with the DB.
-	c.ChanStatus = status
+	c.ChanStatus |= status
```
I think this assumes that `c.ChanStatus == channel.Status` start in the same state. To make sure writes from multiple sources are seen by other `OpenChannel` objects, we should capture the ondisk value to set in memory.
```golang
var chanStatus ChannelStatus
...
        chanStatus = channel.ChannelStatus | status
        channel.ChannelStatus = chanStatus
...
c.ChanStatus = chanStatus
```
Also applies to `MarkDataLoss`

cfromknecht replied:
Since the value of `ChanStatus` is now mutable, we may also need to put it behind a mutex, similar to `ShortChanID()`

halseth replied:
Done!

cfromknecht replied:
nice! in my head i expected that diff to be much bigger, works out nicely that we never set `ChanStatus` manually in tests

cfromknecht commented on lnwallet/channel.go:
```diff
@@ -3149,29 +3156,104 @@ func (lc *LightningChannel) ProcessChanSyncMsg(
+				return nil, nil, nil, err
+			}
+
+			walletLog.Errorf("ChannelPoint(%v), sync failed: " +
```
missing print argument? looks like most of the log statements in this section are missing a chan point

halseth replied:
yeeeh, you are right. Fixed!

cfromknecht commented on lnd_test.go:
```diff
@@ -6001,19 +6050,337 @@ func testRevokedCloseRetributionRemoteHodl(net *lntest.NetworkHarness,
+	// the channel was out of sync, and is now waiting for a force close to
+	// hit the chain.
+	assertNumPendingChannels(t, dave, 1, 0)
+
```
does the test pass if we restart Dave here? this would check Dave's fault tolerance, and we should also assert that the link is not readded to the switch after being marked borked. Can use the `isActive` flag for the latter i think

halseth replied:
Added the restart. 

Not sure if a `ListChannels` gives us any additional coverage here, as we already assert that the number of pending channels is 1. Think that implies that it is not active anymore?

cfromknecht replied:
Yes probably, if we assume that the logic adding the link to the switch on startup always gets the same read of the db as pending channels rpc. I'd say more directly testing the link's activity is worthwhile, even just as a sanity check. Up to you though, not a blocker on the PR

cfromknecht replied:
Better idea: I can make a follow up and sprinkle in these assertions on various closing-related tests :)

cfromknecht commented on channeldb/channel.go:
```diff
@@ -671,6 +688,81 @@ func (c *OpenChannel) MarkAsOpen(openLoc lnwire.ShortChannelID) error {
+	c.Lock()
+	defer c.Unlock()
+
+	if err := c.Db.Update(func(tx *bolt.Tx) error {
```
would be great to use `putChanStatus` here to reduce code duplication, signature might be more like:
```golang
func (c *OpenChannel) putChanStatus(tx *bolt.Tx, status ChannelStatus) (ChannelStatus, error)
```
for easier composition

halseth replied:
Tried doing this, but this turned out to now reduce duplication, since then the DB must be `View`ed from the caller, and the bucket must be retrieved twice in `MarkDataLoss`. The other suggestion were added, PTAL :)

cfromknecht replied:
gotcha, thanks for trying anyway 😂

lcasassa commented:
Nice work here!! Thanks!!

mcduck76 commented:
Sorry for being late to the party, but just today bitcoin lightning wallet refuses to connect with "data loss protection not supported by this peer".
I am running current code base of lnd.

Is there an option to enable the feature? From what I can tell it is on by default since this merge.

Roasbeef commented:
@mcduck76 I'd report that to BLW, looks like they aren't interpreting feature bits correctly.

mcduck76 commented:
> @mcduck76 I'd report that to BLW, looks like they aren't interpreting feature bits correctly.

I can connect from BLW to outher LND nodes. I really can't think of what is wrong with my installation, other nodes can connect fine.
Are there any configuration or make options that I have omitted so there is no support for the feature?
